---
# the default layout is 'page'
icon: fas fa-info-circle
order: 4
---


`scrnaseq2` is an open-source pipeline designed to simplify and automate the analysis of single-cell sequencing data. Hosted on [GitHub](https://github.com/dcgc-bfx/scrnaseq2), this versatile pipeline integrates a series of widely-used bioinformatics tools and [best practices](https://www.nature.com/articles/s41576-023-00586-w), making it easier for researchers to process, analyze, and interpret complex single-cell datasets.

## Key Features

**Automation**: With `scrnaseq2`, users can quickly generate reproducible results by following a streamlined, automated workflow, reducing the need for extensive bioinformatics expertise.

**Customizability**: Although designed to be easy to use, `scrnaseq2` provides extensive customization options for advanced users who want to fine-tune their analyses.

**Documentation**: `scrnaseq2` generates an extensive HTML report that contains additional visualisations, tables and documentation for a better understanding. 

**High scalability**: The workflow is optimized for both small and large datasets, making it suitable for single-cell studies of varying sizes.

**Modular design**: The workflow is flexible and can be adapted to various research needs, supporting multiple stages of single-cell data analysis, including data pre-processing, quality control, normalization, clustering, and marker detection.

**Solid basis**: The workflow is written in R and is largely based on the [Seurat v5](https://satijalab.org/seurat/) R package. 

**Versatility**: The workflow supports single-cell and single-nuclei RNA and protein sequencing data from one or more samples processed with 10X Genomics, SmartSeq, Parse Biosciences and Scale Bio. 

For more information, please visit the [GitHub repository](https://github.com/dcgc-bfx/scrnaseq2).